src/include/detail/linalg/matrix_with_ids.h

```diff
@@ -0,0 +1,209 @@
+  MatrixWithIds& operator=(const MatrixWithIds&) = delete;
+
+  MatrixWithIds(MatrixWithIds&&) = default;
+
+  MatrixWithIds& operator=(MatrixWithIds&& rhs) = default;
+  virtual ~MatrixWithIds() = default;
+
+  MatrixWithIds(
+      Base::size_type nrows,
+      Base::size_type ncols,
+      LayoutPolicy policy = LayoutPolicy()) noexcept
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_right>)
+      : Base(nrows, ncols, policy)
+      , num_ids_(this->num_rows_)
+#ifdef __cpp_lib_smart_ptr_for_overwrite
+      , idsStorage_{std::make_unique_for_overwrite<IdsType[]>(this->num_rows_)}
+#else
+      , idsStorage_{new IdsType[this->num_rows_]}
+#endif
+  {
+  }
+
+  MatrixWithIds(
+      Base::size_type nrows,
+      Base::size_type ncols,
+      LayoutPolicy policy = LayoutPolicy()) noexcept
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_left>)
+      : Base(nrows, ncols, policy)
+      , num_ids_(this->num_cols_)
+#ifdef __cpp_lib_smart_ptr_for_overwrite
+      , idsStorage_{std::make_unique_for_overwrite<IdsType[]>(this->num_cols_)}
+#else
+      , idsStorage_{new IdsType[this->num_cols_]}
+#endif
+  {
+  }
+
+  MatrixWithIds(
+      std::unique_ptr<T[]>&& storage,
+      std::unique_ptr<IdsType[]>&& ids_storage,
+      Base::size_type nrows,
+      Base::size_type ncols,
+      LayoutPolicy policy = LayoutPolicy()) noexcept
+      : Base(std::move(storage), nrows, ncols, policy)
+      , num_ids_{std::is_same<LayoutPolicy, stdx::layout_right>::value ? this->num_rows_ : this->num_cols_}
+      , idsStorage_{std::move(ids_storage)} {
+  }
+
+  /**
+   * Initializer list constructor. Useful for testing and for examples.
+   * The initializer list is assumed to be in row-major order.
+   */
+  MatrixWithIds(
+      std::initializer_list<std::initializer_list<T>> matrix,
+      std::initializer_list<T> ids) noexcept
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_right>)
+      : Base(matrix)
+      , num_ids_(this->num_rows_)
+#ifdef __cpp_lib_smart_ptr_for_overwrite
+      , idsStorage_{std::make_unique_for_overwrite<IdsType[]>(this->num_rows_)}
+#else
+      , idsStorage_{new IdsType[this->num_rows_]}
+#endif
+  {
+    std::copy(ids.begin(), ids.end(), idsStorage_.get());
+  }
+
+  /**
+   * Initializer list constructor. Useful for testing and for examples.
+   * The initializer list is assumed to be in column-major order.
+   */
+  MatrixWithIds(
+      std::initializer_list<std::initializer_list<T>> matrix,
+      std::initializer_list<T> ids) noexcept
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_left>)
+      : Base(matrix)
+      , num_ids_(this->num_cols_)
+#ifdef __cpp_lib_smart_ptr_for_overwrite
+      , idsStorage_{std::make_unique_for_overwrite<IdsType[]>(this->num_cols_)}
+#else
+      , idsStorage_{new IdsType[this->num_cols_]}
+#endif
+  {
+    std::copy(ids.begin(), ids.end(), idsStorage_.get());
+  }
+
+  size_t num_ids() const {
+    return num_ids_;
+  }
+
+  auto ids() {
```

Comment:
When you say `it needs` - why does it _need_ to satisfy these things? From my understanding of how we will use `MatrixWithIds` it does not seem that we _need_ to do this, we only need to be able to get the id at a specific index. And if that is true then going with the current approach will lead to better performance & a more similar call pattern between how we initialize the vectors and the IDs, which will lead to faster and more maintainable code.

Reply:
Talked offline: we do not need to change this for any performance / usage reason. The only ways we access IDs are either to a) get the number of IDs b) access an individual ID. But we agreed to change it because working with `std::vector` could hypothetically be nicer to work with.